sbi/inference/base.py

```diff
@@ -602,16 +615,83 @@ def simulate_for_sbi(
     Returns: Sampled parameters $\theta$ and simulation-outputs $x$.
     """
 
-    theta = proposal.sample((num_simulations,))
+    # 0. if no simulations
+    if num_simulations == 0:
+        theta = torch.tensor([], dtype=float32)
+        x = torch.tensor([], dtype=float32)
+
+    # 0. if some simulations
+    else:
+        # Cast theta to numpy for better joblib performance (seee #1175)
+        seed_all_backends(seed)
+        theta = proposal.sample((num_simulations,)).numpy()
+
+        # 1. if meaningful batching
+        if (
+            simulation_batch_size is not None
+            and simulation_batch_size < num_simulations
```

Comment:
This logic I took from the previous code. To my understanding, supposing that `num_simulations != 0`:
1. **single processing** branches from `simulation_batch_size = 1`
2. **multiprocessing** branches from `1 < simulation_batch_size < num_simulations`
3. **vectorization** (e.g. `numpy` or `torch` broadcasting) branches from
   - `simulation_batch_size == None`
   - `simulation_batch_size > num_simulations`

The change you propose would effectively delete the vectorized branch (which is maybe the quickest way to run the simulations but relies on correctly written simulator code).

We could, in principle, introduce `simulation_batch_size = min(simulation_batch_size, num_simulations)`, but I don't know it this will generate unexpected behaviours in running projects. 
However, I would not go for `if simulation_batch_size is None: simulation_batch_size = 1` since this the `None` case is a valid and meaningful case in the original implementation.

Reply:
Thanks for the structure in this explanation! 

My model of how this should work is that a simulator can be vectorized or not, but that's up the user. If it is not, then the `process_simulator` will have wrapped it to just call a for-loop on the batch. Now, when calling `simulator_for_sbi`, this does not really matter. The `simulation_batch_size` just specified the size of the batches we give to the workers. If the simulator is vectorized, it makes sense to have a large batch size, maybe even `=num_simulations` resulting in a single job. If it is not vectorized, it also makes sense to set it larger than 1 to reduce overhead in setting up the workers I guess. 

Maybe this model was actually not reflected in the current documentation and implementation. 

Moving forward, I suggest to keep the logic that if `simulation_batch_size==None` we set it to `num_simulations`. Same if it is larger than `num_simulations`. Otherwise, we just construct batches and call the workers on them. Thus, the distinction between vectorized and sequential simulator is not made here. 

I am sorry in case my previous comments were not clear on this.

Reply:
Thank you for clarifying! I think I understood now what you mean, we can still use `joblib` to run one big batch with all the data points with minimal overhead and better code readabilty. I proceeded to implement what you suggested.